### DISCUSSION TITLE
Force null or empty values to bottom when sorting?

### DISCUSSION THREAD

**Discussion by sa7mon** [TanStack/table#2371] [General] [answered]
Hi all,

I've got a column in my table that has some empty or null values. Is there an easy way to force these to always be at the bottom no matter which direction the column is sorted? 

This is my custom sort method so far: 
```js
let customSort = (rowA, rowB, columnId) => {
  let a = rowA.values[columnId]
  let b = rowB.values[columnId]

  if (a === '' || a === null) {
    if (b === '' || b === null) {
        return 0; // Both empty/null
    }
    return -1; // Sort a to an index lower than b
  }

  if (b === '' || b === null) {
      if (a === '' || a === null) {
          return 0; // Both empty/null
      }
      return 1; // Sort b to an index lower than a
  }

  if (a > b) return 1;
  if (b > a) return -1;
  return 0;
}
```

which is referenced in my column config: 
```js
...
{
    id: 'applications',
    Header: 'Application ID',
    accessor: a => a.id,
    filterable: true,
    minWidth: 20,
    sortType: customSort,
},
...
```

This seems to work fine when sorting in Descending order, but when I toggle sorting to Ascending, all the blank values are at the top. Is there an easy way to fix this?

Thanks!

**Comment by Zertz** [comment]
Here's what I came up with to keep falsy values at the end:

```js
const sortTypes = useMemo(
    () => ({
        alphanumericFalsyLast(rowA, rowB, columnId, desc) {
            if (!rowA.values[columnId] && !rowB.values[columnId]) {
                return 0;
            }

            if (!rowA.values[columnId]) {
                return desc ? -1 : 1;
            }

            if (!rowB.values[columnId]) {
                return desc ? 1 : -1;
            }

            return rowA.values[columnId].localeCompare(rowB.values[columnId]);
        },
    }),
    []
);

const columns = [{
  sortType: "alphanumericFalsyLast",
}];

useTable({ columns, sortTypes }, useSortBy);
```

**Reply by aruprakshit** [reply]
@Zertz As [MDN doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) said return value may not be always `-1` or `+1`, it can be any positive or negative number, but this lib doc said it must be `+1` or `-1`. So will the user of `localeCompare` as you suggested still work or need a bit more improvement?

**Reply by Zertz** [reply]
Fortunately, that is not the case and there's an [open PR](https://github.com/tannerlinsley/react-table/pull/2820) to clarify the docs.

**Comment by andreyk-code** [comment]
I had the same problem. Got it to work by adding useMemo and using same format as @Zertz. Thanks for sharing!

It's called out in the docs that any custom sort functions have to be memoized. I think this is key to making it work.

**Reply by prasanthbattula** [reply]
```
const sortTypes = React.useMemo(
    () => ({
      alphanumericFalsyLast(rowA, rowB, columnId) {
        
        const a = Number(rowA.values[columnId]);
        const b = Number(rowB.values[columnId]);

        return Number(a) - Number(b);
      },
    }),
    []
  );
```
It worked for me.

**Comment by VincentHokie** [comment]
I had a similar problem today. My solution would translate to something like below for you

```
const customSort = () =>
  React.useCallback((rowA, rowB, columnId, desc) => {

   let a = rowA.values[columnId]
   let b = rowB.values[columnId] 

    /**
     * These first 3 conditions keep our null values at the bottom.
     */
    if (a === null && b !== null) {
      return desc ? -1 : 1;
    }

    if (a !== null && b === null) {
      return desc ? 1 : -1;
    }

    if (a === null && === null) {
      return 0;
    }

    // at this point, you have non-null values and you should do whatever is required to sort those values correctly
    return a -b

  }, []);
```

A more generic answer would be

```
const sortFunctionName = () =>
  React.useCallback((rowA, rowB, columnId, desc) => {

    /**
     * These first 3 conditions keep our null values at the bottom.
     */
    if (conditionRowAToGoBottom && conditionRowBToSortAtTheTop) {
      return desc ? -1 : 1;
    }

    if (conditionRowAToSortAtTheTop && conditionRowBToGoBottom) {
      return desc ? 1 : -1;
    }

    if (conditionRowAToGoBottom && conditionRowBToGoBottom) {
      return 0;
    }

    // at this point, you have values that require sorting at the top of the list, so..do that? :)
    return rowValueA - rowValueB

  }, []);
```

Hope this helps someone

**Comment by lisahillebrand** [comment]
I would like to use a custom sort function and always display rows with undefined value in the respective column at the bottom of the table. They should be at the bottom whether I use desc or asc sorting. The sortingFn does not receive the sort order as param. 

How can this be achieved with v8 without maintaining the sort state myself? 

Thanks for your help.

**Reply by lindesvard** [reply]
Looking for similar options. Since we dont get desc or asc in `sortingFn` its hard to do.

**Comment by ipsips** [comment]
my solution is to translate "empty" values as last unicode character:
```typescript
/**
 * Overriding default sorting function so that empty values are sorted to the end.
 * @see: {@link https://tanstack.com/table/v8/docs/api/features/sorting#sorting-functions}
 */
const LAST_UNICODE_CHARACTER = '\u10FFFF';
sortingFns.basic = (rowA, rowB, columnId) => {
  const valueA = sanitizeValue(rowA.getValue(columnId));
  const valueB = sanitizeValue(rowB.getValue(columnId));
  return valueA > valueB ? 1 : valueA < valueB ? -1 : 0;
};
const sanitizeValue = (v: any) =>
  v === undefined || v === null || v === '' ? LAST_UNICODE_CHARACTER : v;
```

**Comment by KevinVandy** [comment, marked as answer]
`invertSorting` as a column option already exists. https://tanstack.com/table/latest/docs/guide/sorting#sorting-direction

`sortUndefined` as a column option already exists. https://tanstack.com/table/latest/docs/guide/sorting#sort-undefined-or-nullish-values

**Reply by sa7mon** [reply]
Thanks @KevinVandy - I don't think `sortUndefined` existed 4 years ago when I asked this question.

**Reply by KevinVandy** [reply]
This was in response to someone asking about v8 and opening up a similar PR. lol sorry for any confusion

**Reply by ghost** [reply]
@KevinVandy `sortUndefined` does sort `undefined` values, yes, but how about `null` values? Currently `sortUndefined` does not sort `null` values, requiring users to monkey patch the table. 

Here is the code that handles `sortUndefined`: https://github.com/TanStack/table/blob/a4bd09a002949185bd4cca7cd2085faeaf87b682/packages/table-core/src/utils/getSortedRowModel.ts#L60-L77

**Reply by KevinVandy** [reply]
Just have an accessor fallback to undefined instead of null

**Reply by ghost** [reply]
As I said, requiring users to monkey patch the table. Why did the creators decided to just adopt `undefined` and not `null` as well? And are there plans to also include null? Default for missing values when returned from DBs is null.

**Reply by KevinVandy** [reply]
Accessors allow you to control the fallback values for purposes such as this. We can change this for v9 to be `sortNullish`. That makes sense to me, but would technically be a breaking change. The v9 branch has been created recently.

**Reply by donghokimm** [reply]
you can easily make your own `getSortedRowModel` function if needed.

```ts
import {
    getMemoOptions,
    memo,
    type Row,
    type RowData,
    type RowModel,
    type SortingFn,
    type Table,
} from '@tanstack/table-core';

/**
 * @see https://github.com/TanStack/table/blob/db745afdb867511ddd04a79150f89d3894a03532/packages/table-core/src/utils/getSortedRowModel.ts#L5C1-L121C2
 */
export function getSortedRowModel<TData extends RowData>(): (
    table: Table<TData>,
) => () => RowModel<TData> {
    return (table) =>
        memo(
            () => [table.getState().sorting, table.getPreSortedRowModel()],
            (sorting, rowModel) => {
                if (!rowModel.rows.length || !sorting?.length) {
                    return rowModel;
                }

                const sortingState = table.getState().sorting;

                const sortedFlatRows: Row<TData>[] = [];

                // Filter out sortings that correspond to non existing columns
                const availableSorting = sortingState.filter((sort) =>
                    table.getColumn(sort.id)?.getCanSort(),
                );

                const columnInfoById: Record<
                    string,
                    {
                        invertSorting?: boolean;
                        sortingFn: SortingFn<TData>;
                    }
                > = {};

                availableSorting.forEach((sortEntry) => {
                    const column = table.getColumn(sortEntry.id);
                    if (!column) return;

                    columnInfoById[sortEntry.id] = {
                        invertSorting: column.columnDef.invertSorting,
                        sortingFn: column.getSortingFn(),
                    };
                });

                const sortData = (rows: Row<TData>[]) => {
                    // This will also perform a stable sorting using the row index
                    // if needed.
                    const sortedData = rows.map((row) => ({ ...row }));

                    sortedData.sort((rowA, rowB) => {
                        for (let i = 0; i < availableSorting.length; i += 1) {
                            const sortEntry = availableSorting[i]!;
                            const columnInfo = columnInfoById[sortEntry.id]!;
                            const isDesc = sortEntry?.desc ?? false;

                            let sortInt = 0;

                            // All sorting ints should always return in ascending order
                            const aValue = rowA.getValue(sortEntry.id);
                            const bValue = rowB.getValue(sortEntry.id);

                            const aNullable = isNullable(aValue);
                            const bNullable = isNullable(bValue);

                            if (aNullable || bNullable) {
                                return aNullable ? 1 : -1;
                            }

                            if (sortInt === 0) {
                                sortInt = columnInfo.sortingFn(rowA, rowB, sortEntry.id);
                            }

                            // If sorting is non-zero, take care of desc and inversion
                            if (sortInt !== 0) {
                                if (isDesc) {
                                    sortInt *= -1;
                                }

                                if (columnInfo.invertSorting) {
                                    sortInt *= -1;
                                }

                                return sortInt;
                            }
                        }

                        return rowA.index - rowB.index;
                    });

                    // If there are sub-rows, sort them
                    sortedData.forEach((row) => {
                        sortedFlatRows.push(row);
                        if (row.subRows?.length) {
                            row.subRows = sortData(row.subRows);
                        }
                    });

                    return sortedData;
                };

                return {
                    rows: sortData(rowModel.rows),
                    flatRows: sortedFlatRows,
                    rowsById: rowModel.rowsById,
                };
            },
            getMemoOptions(table.options, 'debugTable', 'getSortedRowModel', () =>
                table._autoResetPageIndex(),
            ),
        );
}

function isNullable(value: any) {
    return value === undefined || value === null;
}
```

**Reply by benevbright** [reply]
"easily"?

**Reply by benevbright** [reply]
This is such a bummer. `sortingFn` doesn't also tell me which sorting direction it is, so I can't solve this. Only way is to make all fallback to `undefined`. It's such not necessary to do it... really frustrating.

**Reply by ippeiukai** [reply]
> Accessors allow you to control the fallback values for purposes such as this. We can change this for v9 to be sortNullish. That makes sense to me, but would technically be a breaking change. The v9 branch has been created recently.

It would be nicer if any value can be sorted to beginning or end of list regardless of desc state. It would be even better if that feature could be added in a backward-compatible manner so that we don't have to wait for v9.

Up to v7, sorting functions received `desc` flag, so any value can be put to bottom regardless of asc/desc state using custom sort.
https://react-table-v7-docs.netlify.app/docs/api/useSortBy#column-options

I have a project with many v7 tables that depend on a set of custom sorting functions that sends  empty string always to bottom. As it is too much work to map each table's every such column's values to undefined, this is currently one of the biggest stoppers for us to migrate them to v8. 

*edit (again)*
Found more appropriate issue for this. Moved my idea there.
https://github.com/TanStack/table/issues/6061#issuecomment-4765303706
